src/librustc_trait_selection/traits/project.rs

```diff
@@ -1464,9 +1437,10 @@ fn assoc_ty_def(
         if matches!(item.kind, ty::AssocKind::Type | ty::AssocKind::OpaqueTy)
             && tcx.hygienic_eq(item.ident, assoc_ty_name, trait_def_id)
         {
-            return Ok(specialization_graph::NodeItem {
-                node: specialization_graph::Node::Impl(impl_def_id),
+            return Ok(specialization_graph::LeafDef {
                 item: *item,
+                defining_node: impl_node,
+                finalizing_node: if item.defaultness.is_default() { None } else { Some(impl_node) },
```

Comment:
So I guess this is saying, we'll support a cycle in the inheritance graph construction as long as you have an explicit `type Foo = xxx` declaration, but not if that declaration would've been "inherited" from some other impl? I can see the logic in it, but I wonder if it would be better to draw a sharper line? 

Is this something we have room to change? We should make a not of it for specialization stabilization, it doesn't have to block this PR I don't think.

Reply:
I added a note to the specialization tracking issue about this.